src/Raven.Server/Documents/ResourceCache.cs

```diff
@@ -133,46 +162,56 @@ IEnumerator IEnumerable.GetEnumerator()
 
         public Task<TResource> GetOrAdd(StringSegment databaseName, Task<TResource> task)
         {
-            if (_caseSensitive.TryGetValue(databaseName, out Task<TResource> value))
+            if (_readonlyCaseSensitive.TryGetValue(databaseName, out Task<TResource> value))
                 return value;
 
-            if (_caseInsensitive.TryGetValue(databaseName, out value))
+            if (_readonlyCaseInsensitive.TryGetValue(databaseName, out value))
                 return value;
 
             lock (this)
             {
                 if (_caseInsensitive.TryGetValue(databaseName, out value))
                     return value;
 
-                value = _caseInsensitive.GetOrAdd(databaseName, task);
-                if (value == task)
-                {
-                    _resourceDetails[task] = new ResourceDetails
-                    {
-                        InCacheSince = DateTime.UtcNow
-                    };
-                }
-                _caseSensitive[databaseName] = value;
-                _mappings[databaseName] = new ConcurrentSet<StringSegment>
+                value = GetOrAddInternal(databaseName, task);
+
+                _readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                _readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                _readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
+
+                return value;
+            }
+        }
+
+        private Task<TResource> GetOrAddInternal(StringSegment databaseName, Task<TResource> task)
```

Comment:
Can we not introduce this method? It's just a refactoring and there is no other called of this piece of code. Same motivation as my above comment about `TryRemoveInternal()`

Reply:
See https://github.com/ravendb/ravendb/pull/17942#discussion_r1441632156